File: common/src/main/java/dev/obscuria/fragmentum/v2/api/common/EasingFunction.java
package dev.obscuria.fragmentum.v2.api.common;

import dev.obscuria.fragmentum.v2.core.common._Easing;

@FunctionalInterface
@SuppressWarnings("unused")
public interface EasingFunction {

    float compute(float delta);

    default EasingFunction reverse() {
        return _Easing.reverse(this);
    }

    default EasingFunction scale(float scale) {
        return _Easing.scale(this, scale);
    }

    default EasingFunction merge(EasingFunction other, float ratio) {
        return _Easing.merge(this, other, ratio);
    }

    default EasingFunction mergeOut(EasingFunction other, float ratio) {
        return _Easing.mergeOut(this, other, ratio);
    }
}

File: common/src/main/kotlin/dev/obscuria/fragmentum/api/tools/easing/EasingFunction.kt
package dev.obscuria.fragmentum.api.tools.easing

fun interface EasingFunction {

    fun compute(delta: Float): Float

    fun scale(scale: Float): EasingFunction =
        EasingInternal.scale(this, scale)

    fun reversed(): EasingFunction =
        EasingInternal.reversed(this)

    fun merge(other: EasingFunction, ratio: Float = 0.5f): EasingFunction =
        EasingInternal.merge(this, other, ratio)

    fun mergeOut(other: EasingFunction, ratio: Float = 0.5f): EasingFunction =
        EasingInternal.mergeOut(this, other, ratio)
}